Improve packaging and test coverage

lucaswiman opened the pull request:
@mrocklin This pull request does the following:
## Packaging

Fixes an issue wherein the dependencies need to be installed to even evaluate what the dependencies are. This leads to failures like the following when installing this package in a fresh virtualenv:

``` pytb
(foo)/tmp/foo $ pip install unification
[...]
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "/private/var/folders/0c/wgg0rggj7qn9c5yc1gs5w5lr0000gp/T/pip-build-vwGGU5/unification/setup.py", line 5, in <module>
        import unification
      File "unification/__init__.py", line 1, in <module>
        from .core import unify, reify
      File "unification/core.py", line 3, in <module>
        from toolz.compatibility import iteritems, map
    ImportError: No module named toolz.compatibility
```

This was caused by importing `unification.__version__`, which attempted to import `toolz`.  I fixed it by using a regular expression on the contents of `unification/__init__.py` to determine the version instead of a python import.
## Test Environment

I added a tox.ini file to make running the test suite easier.  Travis is awesome, but it requires a lot of setup to reproduce the tests locally.  I updated `.travis.yml` to use `tox` following [this blog post](https://www.dominicrodger.com/2013/07/26/tox-and-travis/). It's a bit difficult to reproduce issues with it locally [short of running in a docker container](http://stackoverflow.com/questions/21053657/how-to-run-travis-ci-locally), and `tox` seems to be the standard for running paramterized test suites in Python nowadays.
## Test Coverage

I added a few tests to push test coverage up to 100%. I removed the `dicthash` method, since it seems to be unused.
## Thanks!

Thanks for putting this package together! It looks like it'll be really useful!

Commit:
Prevent installation failure when dependencies are not intalled

Commit:
Add tox.ini file and update gitignore

Commit:
Update tests to get test coverage up to 100%

Commit:
Bump version to reflect fix in packaging structure

Commit:
Drop support for python 3.2, since setuptools is kinda broken now.
See https://github.com/python-pillow/pillow-wheels/issues/30 for more details:
> Support for Python 3.0-3.2 has been dropped. Future versions will fail here.
https://bitbucket.org/pypa/setuptools/commits/1cbe02d87fa9b23c3798ef45f3e80e6c75426452

lucaswiman commented:
Build is green now. I dropped python 3.2 support in b05bbcc877d2ff5ce3424d0c0c5262ab0e5b172f, since setuptools seems to no longer be compatible with python 3-3.2.

mrocklin commented on unification/tests/test_more.py:
```diff
@@ -32,6 +34,17 @@ def test_reify_object():
+    class SlotsObject(object):
+        __slots__ = ['myattr']
+        def __init__(self, myattr):
+            self.myattr = myattr
```
Perhaps remove one of the `namedtuple` and the class?

mrocklin replied:
Perhaps an endline after class definition

mrocklin commented on setup.py:
```diff
@@ -1,11 +1,15 @@
+
+contents = re.findall(
+    r'__version__ = \'([.0-9]+)\'',
+    open('unification/__init__.py', 'r').read())[0]
```
This might not work on windows due to the forward slash.

In more recent projects I've just been duplicating the version string in `__init__.py` and `setup.py`.  It's not the most elegant solution but it does work pretty simply.

lucaswiman replied:
Nice catch! One solution would be to use `os.path.join('unification', '__init__.py')` to get a platform-specific directory separator. I'm not very familiar with Python on Windows, but [this stackoverflow thread](http://stackoverflow.com/questions/16333569/mixed-slashes-with-os-path-join-on-windows) suggests it would work. What do you think?

lucaswiman replied:
> In more recent projects I've just been duplicating the version string in **init**.py and setup.py. It's not the most elegant solution but it does work pretty simply.

My experience trying that approach has been that it's really easy to forget the `__init__` version string, though it can work with careful code review. I'm fine with that approach if you don't like the `os.path.join` solution.

mrocklin replied:
Sure.

mrocklin commented:
Hi @lucaswiman thanks for the fixes.  It's great to see someone take up this topic.  What's here looks pretty good I've made a couple of small nitpicky comments.

mrocklin commented on tox.ini:
```diff
@@ -0,0 +1,11 @@
+[tox]
+install_command = pip install {opts} {packages} --use-mirrors
+envlist = py26,py27,py33,py34,pypy
```
Add `py35`?

Commit:
Join paths to avoid hardcoding path separators

Commit:
Clean up SlotsObject test

Commit:
Add py35 to test suite

lucaswiman commented:
Addressed code review comments. Let me know if you want me to rebase the commit history and if you're not comfortable with the solution in 643cfac6270b97706506812e0bb4b070f50b8085 for Windows paths.

mrocklin commented:
Rebasing is up to you.  I'm not actively using this package any longer so I'm not very particular.

lucaswiman commented:
I'm good with leaving the commit history as-is. The build is green, so unless there are more issues, from my view this is good to merge & push to pypi.

>  I'm not actively using this package any longer

I'm curious what the original use case was? I don't think it's quite full-featured enough for my use case yet, so if time allows, I may submit a few more PRs.

mrocklin commented:
I was doing a lot of simplification and rewriting of algebraic expressions in sympy and logpy.

More PRs are welcome.  Assuming things go well I'm also happy to provide commit access and ownership rights.

mrocklin commented:
Pushed to PyPI
